File: .claude/skills/contest-log-collector/adapters/shared/get_github_login.js
// SPDX-License-Identifier: Apache-2.0
// Detect current user's GitHub username via 5-step priority chain.
// Used by OpenCode collector; Claude Code/Codex use the equivalent .py version.
//
// Public API: resolveGithubLogin(repoRoot) → { login, source } | null
// See get_github_login.py header comment for the priority chain (must stay consistent).

import { execFileSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";

const GITHUB_LOGIN_RE = /^[A-Za-z0-9][A-Za-z0-9-]{0,38}$/;

function fromEnv() {
  const val = (process.env.GITHUB_LOGIN || "").trim();
  if (val && GITHUB_LOGIN_RE.test(val)) return { login: val, source: "env" };
  return null;
}

function fromGhHosts() {
  const candidates = [
    join(homedir(), ".config", "gh", "hosts.yml"),
    join(process.env.APPDATA || "", "GitHub CLI", "hosts.yml"),
  ];
  for (const path of candidates) {
    if (!path || !existsSync(path)) continue;
    try {
      const txt = readFileSync(path, "utf8");
      const m = txt.match(/^\s{4}user:\s*(\S+)\s*$/m);
      if (m && GITHUB_LOGIN_RE.test(m[1])) return { login: m[1], source: "gh-hosts" };
    } catch {}
  }
  return null;
}

function fromGhApi() {
  try {
    const out = execFileSync("gh", ["api", "user", "--jq", ".login"], {
      encoding: "utf8", stdio: ["ignore", "pipe", "ignore"], timeout: 3000,
    }).trim();
    if (out && GITHUB_LOGIN_RE.test(out)) return { login: out, source: "gh-api" };
  } catch {}
  return null;
}

function gitRemoteUrl(repoRoot) {
  try {
    return execFileSync("git", ["config", "--get", "remote.origin.url"], {
      cwd: repoRoot, encoding: "utf8", stdio: ["ignore", "pipe", "ignore"], timeout: 3000,
    }).trim() || null;
  } catch { return null; }
}

function fromRemoteBasicAuth(repoRoot) {
  const url = gitRemoteUrl(repoRoot);
  if (!url) return null;
  const m = url.match(/^https?:\/\/([^@:/]+)@github\.com\//);
  if (m && GITHUB_LOGIN_RE.test(m[1])) return { login: m[1], source: "remote-basic-auth" };
  return null;
}

function fromRemoteOwner(repoRoot) {
  const url = gitRemoteUrl(repoRoot);
  if (!url) return null;
  const patterns = [
    /^git@github\.com:([^/]+)\//,
    /^https?:\/\/(?:[^@]+@)?github\.com\/([^/]+)\//,
  ];
  for (const pat of patterns) {
    const m = url.match(pat);
    if (m && GITHUB_LOGIN_RE.test(m[1])) return { login: m[1], source: "remote-owner" };
  }
  return null;
}

function fromNoreplyEmail(repoRoot) {
  try {
    const email = execFileSync("git", ["config", "user.email"], {
      cwd: repoRoot, encoding: "utf8", stdio: ["ignore", "pipe", "ignore"], timeout: 3000,
    }).trim();
    const m = email.match(/^(?:\d+\+)?([^@]+)@users\.noreply\.github\.com$/);
    if (m && GITHUB_LOGIN_RE.test(m[1])) return { login: m[1], source: "noreply-email" };
  } catch {}
  return null;
}

export function resolveGithubLogin(repoRoot) {
  const candidates = [
    fromEnv,
    fromGhHosts,
    fromGhApi,
    () => fromRemoteBasicAuth(repoRoot),
    () => fromRemoteOwner(repoRoot),
    () => fromNoreplyEmail(repoRoot),
  ];
  for (const fn of candidates) {
    const r = fn();
    if (r) return r;
  }
  return null;
}

File: .claude/skills/contest-log-collector/adapters/shared/get_github_login.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
# Detect the current user's GitHub username via a 5-step reliability priority chain.
# Used by Claude Code / Codex collectors (called by snapshot_core.py).
#
# Public API:
#   resolve_github_login(repo_root) -> (login, source) | (None, None)
#
# Priority:
#   0. GITHUB_LOGIN env (explicit override, highest priority)
#   1. ~/.config/gh/hosts.yml active user (offline + authoritative)
#   2. gh api user --jq .login (network + authoritative)
#   3. basic auth user in git remote URL (strong signal)
#   4. git config user.email noreply pattern (last resort)
#
# Deliberately NOT used: git remote owner. The owner of a fork can be an
# organization (e.g. iotpi/foo, open-vela/foo), which is NOT the individual
# contributor's github login and would mis-attribute logs.
#
# Returns (None, None) on failure to let caller fallback.

from __future__ import annotations

import os
import re
import subprocess
from pathlib import Path

GITHUB_LOGIN_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9-]{0,38}$")


def _from_env() -> tuple[str | None, str | None]:
    val = os.environ.get("GITHUB_LOGIN", "").strip()
    if val and GITHUB_LOGIN_RE.match(val):
        return (val, "env")
    return (None, None)


def _from_gh_hosts() -> tuple[str | None, str | None]:
    candidates = [
        Path.home() / ".config" / "gh" / "hosts.yml",
        Path(os.environ.get("APPDATA", "")) / "GitHub CLI" / "hosts.yml",
    ]
    for path in candidates:
        if not path.is_file():
            continue
        try:
            text = path.read_text(encoding="utf-8")
            m = re.search(r"^\s{4}user:\s*(\S+)\s*$", text, re.MULTILINE)
            if m:
                login = m.group(1)
                if GITHUB_LOGIN_RE.match(login):
                    return (login, "gh-hosts")
        except Exception:
            continue
    return (None, None)


def _from_gh_api() -> tuple[str | None, str | None]:
    try:
        r = subprocess.run(
            ["gh", "api", "user", "--jq", ".login"],
            capture_output=True, text=True, timeout=3,
        )
        if r.returncode == 0:
            login = r.stdout.strip()
            if login and GITHUB_LOGIN_RE.match(login):
                return (login, "gh-api")
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass
    return (None, None)


def _git_remote_url(repo_root: Path) -> str | None:
    try:
        r = subprocess.run(
            ["git", "config", "--get", "remote.origin.url"],
            cwd=str(repo_root), capture_output=True, text=True, timeout=3,
        )
        if r.returncode == 0:
            return r.stdout.strip() or None
    except Exception:
        pass
    return None


def _from_remote_basic_auth(repo_root: Path) -> tuple[str | None, str | None]:
    url = _git_remote_url(repo_root)
    if not url:
        return (None, None)
    m = re.match(r"^https?://([^@:/]+)@github\.com/", url)
    if m:
        login = m.group(1)
        if GITHUB_LOGIN_RE.match(login):
            return (login, "remote-basic-auth")
    return (None, None)


def _from_noreply_email(repo_root: Path) -> tuple[str | None, str | None]:
    try:
        r = subprocess.run(
            ["git", "config", "user.email"],
            cwd=str(repo_root), capture_output=True, text=True, timeout=3,
        )
        if r.returncode != 0:
            return (None, None)
        email = r.stdout.strip()
        m = re.match(r"^(?:\d+\+)?([^@]+)@users\.noreply\.github\.com$", email)
        if m:
            login = m.group(1)
            if GITHUB_LOGIN_RE.match(login):
                return (login, "noreply-email")
    except Exception:
        pass
    return (None, None)


def resolve_github_login(repo_root: Path) -> tuple[str | None, str | None]:
    for fn in (
        _from_env,
        _from_gh_hosts,
        _from_gh_api,
        lambda: _from_remote_basic_auth(repo_root),
        lambda: _from_noreply_email(repo_root),
    ):
        login, source = fn()
        if login:
            return (login, source)
    return (None, None)


if __name__ == "__main__":
    import sys
    target = Path(sys.argv[1]) if len(sys.argv) > 1 else Path.cwd()
    login, source = resolve_github_login(target)
    if login:
        print(f"{login}\t{source}")
        sys.exit(0)
    print("ERROR: cannot detect GitHub login. Set GITHUB_LOGIN env var.", file=sys.stderr)
    sys.exit(1)
